### ISSUE TITLE
Thread-unsafe issue if we set ack=True in SubscriptionListener

### ISSUE THREAD

**Issue by LiyiW** [nikipore/stompest#51]
In my case, I set ack=True (which is also the default value) when create a stompest.async.listener.SubscriptionListener instance. And the onMessage function will call connection.ack(frame) after message handler finishes(**stompest.async.listener.py line 220**). But thread-unsafe issue comes up if I set the twisted reactor to contains more threads, like reactor.suggestThreadPoolSize(size=4) .

The reason is **yield** connection.ack(frame) can cause the ack action to be executed in another thread different from main thread. If you trace the ack action, this will ultimately twisted.internet.abstract.FileDescriptor.write(self, data). This function is thread-unsafe because some data is corrupted(**twisted.internet.abstract.py line 357**).

The result is, some of the acks are missing in fact although there is log of "Sending ACK frame...".

Now, I set ack=False and do ack in main thread to solve this problem

**Comment by LiyiW** [comment]
You are right, only by telling the reactor to do run in another thread can that happen.
I provide an example.
[twisted_test_consumer.txt](https://github.com/nikipore/stompest/files/1790052/twisted_test_consumer.txt)
[twisted_test_producer.txt](https://github.com/nikipore/stompest/files/1790053/twisted_test_producer.txt)

Just change the suffix of the files to  be .py and run them.
As you will see, the consumer logs "Sending ACK frame..." in a thread different from main thread.

In my message handler "consume", I tell the reactor to run "callback_func" in another thread because somehow it runs for a while and I hope multiple messages be processed in parallel. As a result, the rest of the "onMessage" function(stompest/src/async/stompest/async/listener.py  line 207) will also run in that thread due to the mechanism of Deferred and reactor.

As we can't let ack run in main thread again, I suggest a lock be acquired at before telling twisted transport to write, just use lock in the StompProtocol object 

in __init__:
self._lock = threading.RLock()

and do the following
    with self._lock:
            self.transport.write(binaryType(frame)) 
(stompest/src/async/stompest/async/protocol.py line 48)  

I think this resolves the problem

**Comment by LiyiW** [comment]
Hi, Jan

I replied to you on github with an example and solution. Hope that solves the problem.

Best!
Liyi

发件人： Jan Müller
发送时间： 2018-03-06 13:35:36
收件人：  nikipore/stompest
抄送：  LiyiW; Author
主题： Re: [nikipore/stompest] Thread-unsafe issue if we set ack=True inSubscriptionListener (#51)
Thanks for this heads-up. Any idea as to how to fix this?

**Comment by nikipore** [comment]
Synchronizing the access to the transport level is exactly what you shouldn't do in an asynchronous (sic!) setting, so your suggested solution is invalid.

Looking at your message handler
```
@defer.inlineCallbacks
def consume(self, client, frame):
    d = defer.Deferred()
    d.addCallback(longRunningCalculation)
    reactor.callInThread(d.callback, "result 1")
    yield d
```
You are doing something very wicked, namely calling a callback from another thread, so you `yield` from one (the main) thread and `send` from another, which messes up the coroutine and imho breaks the reactor (which *must* remain single-threaded) altogether. If you do it this way (untested), it should work:
```
from twisted.internet import threads
[...]
@defer.inlineCallbacks
def consume(self, client, frame):
    result = yield threads.deferToThread(longRunningCalculation)
    logging.debug("In the main thread: %s", result)
```
